sbnci/Modules/RecoEff/RecoEff_module.cc

```diff
@@ -0,0 +1,417 @@
+// File:         RecoEff_module.cc                                       //
+// Author:       Henry Lay h.lay@lancaster.ac.uk                         //
+//                                                                       //
+// Purpose:      The module saves truth information regarding the        //
+//               primary products of neutrino interactions. It also      //
+//               saves their status in reconstruction. This allows       //
+//               for plotting of a series of plots of different          //
+//               reconstruction metrics.                                 //
+///////////////////////////////////////////////////////////////////////////
+
+//Standard
+#include "art/Framework/Core/EDAnalyzer.h"
+#include "art/Framework/Core/ModuleMacros.h"
+#include "art/Framework/Principal/Event.h"
+#include "art/Framework/Principal/Handle.h"
+#include "art/Framework/Principal/Run.h"
+#include "art/Framework/Principal/SubRun.h"
+#include "canvas/Utilities/InputTag.h"
+#include "fhiclcpp/ParameterSet.h"
+#include "messagefacility/MessageLogger/MessageLogger.h"
+
+//Sim Base
+#include "nusimdata/SimulationBase/MCParticle.h"
+#include "nusimdata/SimulationBase/MCTruth.h"
+
+//Reco Base
+#include "lardataobj/RecoBase/PFParticle.h"
+#include "lardataobj/RecoBase/Track.h"
+#include "lardataobj/RecoBase/Shower.h"
+#include "lardataobj/RecoBase/Hit.h"
+
+//Tools
+#include "canvas/Persistency/Common/FindManyP.h"
+#include "canvas/Persistency/Common/FindOneP.h"
+
+//LArSoft
+#include "larsim/Utils/TruthMatchUtils.h"
+#include "lardata/DetectorInfoServices/DetectorClocksService.h"
+
+//Root
+#include "art_root_io/TFileService.h"
+#include "TTree.h"
+
+class RecoEff;
+
+class RecoEff : public art::EDAnalyzer {
+public:
+  explicit RecoEff(fhicl::ParameterSet const& pset);
+
+  RecoEff(RecoEff const&) = delete;
+  RecoEff(RecoEff&&) = delete;
+  RecoEff& operator=(RecoEff const&) = delete;
+  RecoEff& operator=(RecoEff&&) = delete;
+
+  void analyze(art::Event const &e) override;
+
+private:
+
+  void ClearData();
+  void SetupHitsMap();
+  void ReconstructionProcessor(art::Event const &e);
+  void TruthProcessor(art::Event const &e);
+
+  std::vector<art::Ptr<recob::PFParticle> > GetPrimaryPFPs();
+  art::Ptr<recob::PFParticle> GetPFP(long unsigned int const &id);
+
+  float Purity(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID);
+  float Completeness(std::vector< art::Ptr<recob::Hit> > const &objectHits, int const &trackID);
+  float TrackLength(art::Ptr<recob::Track> const &track);
+  float TrueTrackLength(art::Ptr<simb::MCParticle> const &particle);
+
+
+  art::Handle<std::vector<simb::MCTruth> > eHandleNeutrinos;
+  art::Handle<std::vector<recob::Hit> > eHandleHits;
+  art::Handle<std::vector<recob::Track> > eHandleTracks;
+  art::Handle<std::vector<recob::Shower> > eHandleShowers;
+  art::Handle<std::vector<recob::PFParticle> > eHandlePFPs;
```

Comment:
Do you need to keep these as data members.  Handles become very meaningless between events  and you risk using corrupt data if not cleared properly

Reply:
Have changed to creating within each event in [this](https://github.com/SBNSoftware/sbnci/pull/7/commits/f5f93b9c40fbac2b9f741b58ceaa3b543f486305) commit.